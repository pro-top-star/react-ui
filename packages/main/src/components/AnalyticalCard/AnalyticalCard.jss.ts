import * as ThemingParameters from '@ui5/webcomponents-react-base/lib/sap_fiori_3';
import * as spacing from '@ui5/webcomponents-react-base/lib/spacing';

const styles = {
  card: {
    backgroundColor: ThemingParameters.sapUiTileBackground,
    // TODO There is a border mentioned in the specs, but this one looks weird.
    // border: `0.625rem solid ${ThemingParameters.sapUiTileBorderColor}`,
    boxShadow: ThemingParameters.sapUiShadowLevel0,
    borderRadius: '0.25rem',
    textAlign: 'start',
    overflow: 'hidden',
    position: 'relative',
    fontFamily: ThemingParameters.sapUiFontFamily,
    boxSizing: 'border-box'
  },
  content: spacing.sapUiContentPadding
};

export default styles;
