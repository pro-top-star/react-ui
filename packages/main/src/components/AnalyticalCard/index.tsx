import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import React, { CSSProperties, FC, forwardRef, ReactNode, ReactNodeArray, Ref, useMemo } from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { JSSTheme } from '../../interfaces/JSSTheme';

import styles from './AnalyticalCard.jss';

export interface AnalyticalCardTypes extends CommonProps {
  /**
   * The Card header Component, using the AnalyticalCardHeader is recommended.
   */
  header?: ReactNode;
  /**
   * Expected one or more React Components
   */
  children: ReactNode | ReactNodeArray;
  width?: CSSProperties['width'];
}

const useStyles = createUseStyles<JSSTheme, keyof ReturnType<typeof styles>>(styles, { name: 'AnalyticalCard' });

export const AnalyticalCard: FC<AnalyticalCardTypes> = forwardRef(
  (props: AnalyticalCardTypes, ref: Ref<HTMLDivElement>) => {
    const { children, style, className, tooltip, header, width } = props;
    const classes = useStyles();
    const classNameString = StyleClassHelper.of(classes.card);
    if (className) {
      classNameString.put(className);
    }

    const analyticalCardStyles = useMemo(() => {
      return {
        width,
        ...style
      };
    }, [style, width]);

    const passThroughProps = usePassThroughHtmlProps(props);

    return (
      <div
        ref={ref}
        className={classNameString.toString()}
        style={analyticalCardStyles}
        title={tooltip}
        {...passThroughProps}
      >
        {header}
        <div className={classes.content}>{children}</div>
      </div>
    );
  }
);

AnalyticalCard.displayName = 'AnalyticalCard';
AnalyticalCard.defaultProps = { width: '20rem', header: null };
