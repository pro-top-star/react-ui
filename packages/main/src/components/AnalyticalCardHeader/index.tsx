import { Event } from '@ui5/webcomponents-react-base/lib/Event';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { DeviationIndicator } from '@ui5/webcomponents-react/lib/DeviationIndicator';
import { FlexBox } from '@ui5/webcomponents-react/lib/FlexBox';
import { FlexBoxAlignItems } from '@ui5/webcomponents-react/lib/FlexBoxAlignItems';
import { FlexBoxDirection } from '@ui5/webcomponents-react/lib/FlexBoxDirection';
import { FlexBoxJustifyContent } from '@ui5/webcomponents-react/lib/FlexBoxJustifyContent';
import { FlexBoxWrap } from '@ui5/webcomponents-react/lib/FlexBoxWrap';
import { ObjectStatus } from '@ui5/webcomponents-react/lib/ObjectStatus';
import { ValueState } from '@ui5/webcomponents-react/lib/ValueState';
import React, { FC, forwardRef, Ref, useCallback, useMemo } from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { JSSTheme } from '../../interfaces/JSSTheme';
import styles from './AnalyticalCardHeader.jss';

export interface AnalyticalCardHeaderPropTypes extends CommonProps {
  title?: string;
  subTitle?: string;
  arrowIndicator?: DeviationIndicator;
  showIndicator?: boolean;
  indicatorState?: ValueState;
  value?: string;
  unit?: string;
  valueState?: ValueState;
  target?: string;
  deviation?: string;
  onHeaderPress?: (event: Event) => void;
  description?: string;
  counter?: string;
  counterState?: ValueState;
  currency?: string;
}

const useStyles = createUseStyles<JSSTheme, keyof ReturnType<typeof styles>>(styles, {
  name: 'AnalyticalCardHeader'
});

export const AnalyticalCardHeader: FC<AnalyticalCardHeaderPropTypes> = forwardRef(
  (props: AnalyticalCardHeaderPropTypes, ref: Ref<HTMLDivElement>) => {
    const {
      title,
      subTitle,
      value,
      unit,
      target,
      deviation,
      valueState,
      onHeaderPress,
      showIndicator,
      tooltip,
      className,
      description,
      counter,
      counterState,
      currency,
      indicatorState,
      arrowIndicator,
      style
    } = props;
    const classes = useStyles(props);
    const onClick = useCallback(
      (e) => {
        if (onHeaderPress) {
          onHeaderPress(Event.of(null, e));
        }
      },
      [onHeaderPress]
    );
    const indicatorIcon = useMemo(() => {
      const arrowClasses = StyleClassHelper.of(classes.arrowIndicatorShape);
      switch (arrowIndicator) {
        case DeviationIndicator.Up:
          arrowClasses.put(classes.arrowUp);
          break;
        case DeviationIndicator.Down:
          arrowClasses.put(classes.arrowDown);
          break;
        default:
          arrowClasses.put(classes.arrowRight);
          break;
      }

      switch (indicatorState) {
        case ValueState.Success:
          arrowClasses.put(classes.good);
          break;
        case ValueState.Error:
          arrowClasses.put(classes.error);
          break;
        case ValueState.Warning:
          arrowClasses.put(classes.critical);
          break;
        default:
          arrowClasses.put(classes.none);

          break;
      }
      return <div className={arrowClasses.valueOf()} />;
    }, [arrowIndicator, indicatorState, classes]);

    const headerClasses = StyleClassHelper.of(classes.cardHeader);
    if (onHeaderPress) {
      headerClasses.put(classes.cardHeaderClickable);
    }

    const valueAndUnitClasses = StyleClassHelper.of(classes.valueAndUnit);
    if (valueState === ValueState.Error) {
      valueAndUnitClasses.put(classes.error);
    }
    if (valueState === ValueState.Warning) {
      valueAndUnitClasses.put(classes.critical);
    }
    if (valueState === ValueState.Success) {
      valueAndUnitClasses.put(classes.good);
    }

    if (className) {
      headerClasses.put(className);
    }
    const shouldRenderContent = [value, unit, deviation, target].some((v) => v !== null);

    const passThroughProps = usePassThroughHtmlProps(props);

    return (
      <div
        ref={ref}
        onClick={onClick}
        className={headerClasses.valueOf()}
        title={tooltip}
        style={style}
        {...passThroughProps}
      >
        <div className={classes.headerContent}>
          <div className={classes.headerTitles}>
            <FlexBox justifyContent={FlexBoxJustifyContent.SpaceBetween} wrap={FlexBoxWrap.NoWrap}>
              <div className={classes.headerText}>{title}</div>
              <ObjectStatus className={classes.counter} state={counterState}>
                {counter}
              </ObjectStatus>
            </FlexBox>
            <div className={classes.subHeaderText}>
              {subTitle}
              {currency && ` | ${currency}`}
            </div>
          </div>
          {shouldRenderContent && (
            <FlexBox direction={FlexBoxDirection.Row} className={classes.kpiContent} alignItems={FlexBoxAlignItems.End}>
              <FlexBox direction={FlexBoxDirection.Row}>
                <div className={valueAndUnitClasses.valueOf()}>
                  <div className={classes.value}>{value}</div>
                  <div className={classes.indicatorAndUnit}>
                    {showIndicator && indicatorIcon}
                    <div className={classes.unit}>{unit}</div>
                  </div>
                </div>
              </FlexBox>
              <FlexBox
                direction={FlexBoxDirection.Row}
                wrap={FlexBoxWrap.NoWrap}
                className={classes.targetAndDeviation}
              >
                {target !== null && (
                  <FlexBox
                    direction={FlexBoxDirection.Column}
                    className={classes.targetAndDeviationColumn}
                    wrap={FlexBoxWrap.NoWrap}
                  >
                    <span>Target</span>
                    <span className={classes.targetAndDeviationValue}>{target}</span>
                  </FlexBox>
                )}
                {deviation !== null && (
                  <FlexBox
                    direction={FlexBoxDirection.Column}
                    className={classes.targetAndDeviationColumn}
                    wrap={FlexBoxWrap.NoWrap}
                  >
                    <span>Deviation</span>
                    <span className={classes.targetAndDeviationValue}>{deviation}</span>
                  </FlexBox>
                )}
              </FlexBox>
            </FlexBox>
          )}
          <div className={classes.description}>{description}</div>
        </div>
      </div>
    );
  }
);

AnalyticalCardHeader.displayName = 'AnalyticalCardHeader';

AnalyticalCardHeader.defaultProps = {
  title: null,
  subTitle: null,
  arrowIndicator: DeviationIndicator.None,
  showIndicator: true,
  indicatorState: ValueState.None,
  value: null,
  unit: null,
  valueState: ValueState.None,
  target: null,
  deviation: null,
  onHeaderPress: null,
  description: null,
  counter: null,
  counterState: ValueState.None,
  currency: null
};
