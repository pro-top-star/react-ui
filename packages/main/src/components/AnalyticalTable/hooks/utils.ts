export const makeTemplateColumns = (columns, resizedColumns) => {
  return columns
    .map((column) => {
      if (resizedColumns[column.id]) {
        return `${resizedColumns[column.id]}px`;
      }
      return `minmax(${column.minWidth}px, ${column.width || '1fr'})`;
    })
    .join(' ');
};
