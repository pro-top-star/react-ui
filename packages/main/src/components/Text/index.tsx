import { createComponentStyles } from '@ui5/webcomponents-react-base/lib/createComponentStyles';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import React, { FC, forwardRef, ReactNode, Ref } from 'react';
import { CommonProps } from '../../interfaces/CommonProps';
import { TextStyles } from './Text.jss';

export interface TextProps extends CommonProps {
  /**
   * Pass the text as direct child of Text
   */
  children: string | JSX.Element | ReactNode;

  renderWhitespace?: boolean;

  wrapping?: boolean;
}

const useStyles = createComponentStyles(TextStyles, { name: 'Text' });

/**
 * <code>import { Text } from '@ui5/webcomponents-react/lib/Text';</code>
 */
const Text: FC<TextProps> = forwardRef((props: TextProps, ref: Ref<HTMLSpanElement>) => {
  const { children, renderWhitespace, wrapping, className, style, tooltip, slot } = props;
  const classes = useStyles();
  const classNameString = StyleClassHelper.of(classes.text);
  if (wrapping === false) {
    classNameString.put(classes.noWrap);
  }
  if (renderWhitespace) {
    classNameString.put(classes.renderWhitespace);
  }
  if (className) {
    classNameString.put(className);
  }

  const passThroughProps = usePassThroughHtmlProps(props);

  return (
    <span
      ref={ref}
      style={style}
      className={classNameString.toString()}
      title={tooltip}
      slot={slot}
      {...passThroughProps}
    >
      {children}
    </span>
  );
});

Text.defaultProps = {
  renderWhitespace: false,
  wrapping: true
};

Text.displayName = 'Text';

export { Text };
