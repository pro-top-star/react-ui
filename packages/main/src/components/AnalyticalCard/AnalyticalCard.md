```jsx
import { AnalyticalCard } from '@ui5/webcomponents-react/lib/AnalyticalCard';
```
