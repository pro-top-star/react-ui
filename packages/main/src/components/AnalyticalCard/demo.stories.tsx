import { action } from '@storybook/addon-actions';
import { boolean, select, text } from '@storybook/addon-knobs';
import { LineChart } from '@ui5/webcomponents-react-charts/lib/LineChart';
import { AnalyticalCard } from '@ui5/webcomponents-react/lib/AnalyticalCard';
import { AnalyticalCardHeader } from '@ui5/webcomponents-react/lib/AnalyticalCardHeader';
import { DeviationIndicator } from '@ui5/webcomponents-react/lib/DeviationIndicator';
import { ValueState } from '@ui5/webcomponents-react/lib/ValueState';
import React from 'react';
import notes from './AnalyticalCard.md';

export default {
  title: 'Components / Analytical Card',
  component: AnalyticalCard,
  parameters: { notes }
};

export const defaultStory = () => (
  <AnalyticalCard
    width={text('width', '20rem')}
    header={
      <AnalyticalCardHeader
        title={text('title', 'Title')}
        subTitle={text('subTitle', 'Subtitle')}
        arrowIndicator={select('arrowIndicator', DeviationIndicator, DeviationIndicator.Down)}
        indicatorState={select('indicatorState', ValueState, ValueState.Success)}
        value={text('value', 'Value')}
        valueState={select('valueState', ValueState, ValueState.Success)}
        unit={text('unit', 'Unit')}
        target={text('target', 'target')}
        deviation={text('deviation', 'deviation')}
        onHeaderPress={action('Header pressed')}
        showIndicator={boolean('showIndicator', true)}
        description={text('description', 'Description')}
        counter={text('counter', 'Counter')}
        counterState={select('counterState', ValueState, ValueState.Success)}
        currency={text('currency', 'EUR')}
      />
    }
  >
    <LineChart
      labels={[
        '201901',
        '201902',
        '201903',
        '201904',
        '201905',
        '201906',
        '201907',
        '201908',
        '201909',
        '201910',
        '201911',
        '201912'
      ]}
      datasets={[
        {
          label: 'automated mapping quota',
          data: [65, 64, 68, 73, 75, 75, 79, 84, 88, 88.5, 89, 92]
        }
      ]}
      height={200}
      width={270}
      options={{
        layout: {
          padding: { top: 25 }
        }
      }}
      valueAxisFormatter={(yValue) => {
        return `${yValue}%`;
      }}
      colors={['#607182']}
    />
  </AnalyticalCard>
);

defaultStory.story = { name: 'Default' };
