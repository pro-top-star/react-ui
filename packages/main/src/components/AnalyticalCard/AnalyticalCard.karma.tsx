import { mountThemedComponent } from '@shared/tests/utils';
import { expect, use } from 'chai';
import { matchSnapshot } from 'chai-karma-snapshot';
import React from 'react';
import { AnalyticalCard } from '../../lib/AnalyticalCard';
import { AnalyticalCardHeader } from '../../lib/AnalyticalCardHeader';
import { DeviationIndicator } from '../../lib/DeviationIndicator';
import { Text } from '../../lib/Text';
import { ValueState } from '../../lib/ValueState';

use(matchSnapshot);

const Header = (
  <AnalyticalCardHeader
    title="Title"
    subTitle="Subtitle"
    arrowIndicator={DeviationIndicator.Down}
    indicatorState={ValueState.Success}
    value="Value"
    valueState={ValueState.Success}
    unit="Unit"
    target="Target"
    deviation="Deviation"
    showIndicator
    description="Description"
  />
);

// TODO Add more tests
describe('Analytical Card', () => {
  it('Render without Crashing', () => {
    const wrapper = mountThemedComponent(
      <AnalyticalCard header={Header}>
        <Text>I'm a content!</Text>
      </AnalyticalCard>
    );
    expect(wrapper.debug()).to.matchSnapshot();
  });
});
