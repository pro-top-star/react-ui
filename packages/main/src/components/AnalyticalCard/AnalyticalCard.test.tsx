import { createPassThroughPropsTest, mountThemedComponent } from '@shared/tests/utils';
import React from 'react';
import { AnalyticalCard } from '@ui5/webcomponents-react/lib/AnalyticalCard';
import { AnalyticalCardHeader } from '@ui5/webcomponents-react/lib/AnalyticalCardHeader';
import { DeviationIndicator } from '@ui5/webcomponents-react/lib/DeviationIndicator';
import { Text } from '@ui5/webcomponents-react/lib/Text';
import { ValueState } from '@ui5/webcomponents-react/lib/ValueState';

const Header = (
  <AnalyticalCardHeader
    title="Title"
    subTitle="Subtitle"
    arrowIndicator={DeviationIndicator.Down}
    indicatorState={ValueState.Success}
    value="Value"
    valueState={ValueState.Success}
    unit="Unit"
    target="Target"
    deviation="Deviation"
    showIndicator
    description="Description"
  />
);

// TODO Add more tests
describe('Analytical Card', () => {
  test('Render without Crashing', () => {
    const wrapper = mountThemedComponent(
      <AnalyticalCard header={Header}>
        <Text>I'm a content!</Text>
      </AnalyticalCard>
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  createPassThroughPropsTest(AnalyticalCard);
});
