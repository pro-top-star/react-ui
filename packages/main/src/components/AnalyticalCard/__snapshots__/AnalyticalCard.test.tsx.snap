// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytical Card Render without Crashing 1`] = `
<div
  class="AnalyticalCard-card-0"
>
  <div
    class="AnalyticalCardHeader-cardHeader-0"
  >
    <div
      class="AnalyticalCardHeader-headerContent-0"
    >
      <div
        class="AnalyticalCardHeader-headerTitles-0"
      >
        <div
          class="FlexBox-flexBox-0 FlexBox-flexBoxDirectionRow-0 FlexBox-justifyContentSpaceBetween-0 FlexBox-alignItemsStretch-0 FlexBox-flexWrapNoWrap-0"
        >
          <div
            class="AnalyticalCardHeader-headerText-0"
          >
            Title
          </div>
          <div
            class="ObjectStatus-objectStatus-0 AnalyticalCardHeader-counter-0"
          />
        </div>
        <div
          class="AnalyticalCardHeader-subHeaderText-0"
        >
          Subtitle
        </div>
      </div>
      <div
        class="FlexBox-flexBox-0 FlexBox-flexBoxDirectionRow-0 FlexBox-justifyContentStart-0 FlexBox-alignItemsEnd-0 FlexBox-flexWrapNoWrap-0 AnalyticalCardHeader-kpiContent-0"
      >
        <div
          class="FlexBox-flexBox-0 FlexBox-flexBoxDirectionRow-0 FlexBox-justifyContentStart-0 FlexBox-alignItemsStretch-0 FlexBox-flexWrapNoWrap-0"
        >
          <div
            class="AnalyticalCardHeader-valueAndUnit-0 AnalyticalCardHeader-good-0"
          >
            <div
              class="AnalyticalCardHeader-value-0"
            >
              Value
            </div>
            <div
              class="AnalyticalCardHeader-indicatorAndUnit-0"
            >
              <div
                class="AnalyticalCardHeader-arrowIndicatorShape-0 AnalyticalCardHeader-arrowDown-0 AnalyticalCardHeader-good-0"
              />
              <div
                class="AnalyticalCardHeader-unit-0"
              >
                Unit
              </div>
            </div>
          </div>
        </div>
        <div
          class="FlexBox-flexBox-0 FlexBox-flexBoxDirectionRow-0 FlexBox-justifyContentStart-0 FlexBox-alignItemsStretch-0 FlexBox-flexWrapNoWrap-0 AnalyticalCardHeader-targetAndDeviation-0"
        >
          <div
            class="FlexBox-flexBox-0 FlexBox-flexBoxDirectionColumn-0 FlexBox-justifyContentStart-0 FlexBox-alignItemsStretch-0 FlexBox-flexWrapNoWrap-0 AnalyticalCardHeader-targetAndDeviationColumn-0"
          >
            <span>
              Target
            </span>
            <span
              class="AnalyticalCardHeader-targetAndDeviationValue-0"
            >
              Target
            </span>
          </div>
          <div
            class="FlexBox-flexBox-0 FlexBox-flexBoxDirectionColumn-0 FlexBox-justifyContentStart-0 FlexBox-alignItemsStretch-0 FlexBox-flexWrapNoWrap-0 AnalyticalCardHeader-targetAndDeviationColumn-0"
          >
            <span>
              Deviation
            </span>
            <span
              class="AnalyticalCardHeader-targetAndDeviationValue-0"
            >
              Deviation
            </span>
          </div>
        </div>
      </div>
      <div
        class="AnalyticalCardHeader-description-0"
      >
        Description
      </div>
    </div>
  </div>
  <div
    class="AnalyticalCard-content-0"
  >
    <span
      class="Text-text-0"
    >
      I'm a content!
    </span>
  </div>
</div>
`;
